### ISSUE TITLE
Can I create metric(accuracy)?

### ISSUE THREAD

**Issue by SeungJunJang** [liznerski/fcdd#49]
First of all, thank you to the developer.

When measuring model performance, I want to use accuracy as well as 'AUC'.
I'm asking you a question because I feel so hard when I try to modify the code.

I try to modify this function( [line 328](https://github.com/liznerski/fcdd/blob/master/python/fcdd/training/bases.py#L328) )and use it.

I understand that the algorithm uses existing labels and score(red_ascores) to obtain AUC.
loss is same anomaly score(pseudo-Huber loss).
anomaly score is loss. ( ascores == anomaly score)
Find tpr, fpr, and thresholds through the "roc_auc_curve" function &
the reduce_ascore(""" Reduces the anomaly score to be a score per image (detection). """) through the "roc_auc_score" function.

However, it is difficult to obtain 'accuracy' in this way.
I understand that "anomaly score" is only used to 'explain' through heat maps.

Ask how to configure the algorithm of 'accuracy'.
I think I need a anomaly detection(binary classification) result for the test dataset.
How do I implement this part?

**Comment by liznerski** [comment]
You're welcome :)

As you explained in #46, you want to perform multi-class classification with a one vs. rest approach. For computing the full multi-class accuracy, you'll need to change quite a bit as the current code always treats everything as an anomaly detection problem. I'd probably write a completely new script that loads the snapshots from the trained FCDD AD models and applies them to multi-class classification.

If you want to compute a **binary** accuracy, you'll need to define a threshold that turns real anomaly scores to binary decisions. For example, you could set the threshold to 5, which turns every anomaly scores below 5 to 0 (i.e., normal) and every one above to 1 (i.e., anomalous). The intuition behind the AUC metric is that it provides a threshold-independent evaluation by "averaging" over all possible thresholds. Thus, we don't have an implementation for finding a threshold. Since you want to perform classification in the end, you have ground-truth "outliers" (samples from other classes). You can define a validation split of the data and use these ground-truth samples together with some hold-out normal samples to find the best threshold empirically. With that, you can then compute the accuracy easily.

Did that help?

**Comment by SeungJunJang** [comment]
Thank you for your answer.

I also thought that it was possible only by utilizing thresholds, and I was thinking about a non-empirical way of finding them.

I have two questions.

1. If I want to make it 0 and 1 through the threshold, do you think it is appropriate to make it a scalar value in the ||A||_1 method of calculating the output matrix for each image?

2. I'd like to ask you if you have any ideas on how to find a non-empirical threshold.

**Comment by liznerski** [comment]
I think these question become out of scope for this repository. Feel free to write me a mail to engage further discussion :)

**Comment by SeungJunJang** [comment]
Then I'll ask you another question.
My code

>  sudo python runners/run_custom.py --logdir /mnt/c/JSJ/FCDD_1020_Code_Check1 --datadir /mnt/c/JSJ/Casting_Product --objective fcdd -b 400 -e 1 --it 1 -d custom -n FCDD_CNN32_LW3K --preproc none --supervise-mode noise --noise-mode cifar100 --nominal-label 0 --blur-heatmaps --acc-batches 1 --oe-limit 15

My data is

> -custom 
> 
>  -  train
>     - class1 
>       - [normal ] : 2875 (number of data)
>       - [anomalous] : 3758 (number of data)
>  -  test
>     - class1 
>       - [normal ] : 262 (number of data)
>       - [anomalous]  : 453 (number of data)

**> Q1. I'm asking because I saw a strange part of the number of labels in the data in the process of modifying the code.**


The exact number of data is shown at the output in this picture.
[line 104](https://github.com/liznerski/fcdd/blob/master/python/fcdd/training/setup.py#L104) 


![training_setup_dataset](https://user-images.githubusercontent.com/35719368/196882532-60b3d00c-9202-470d-a650-604d7a25d736.png)

But,

[line 61](https://github.com/liznerski/fcdd/blob/master/python/fcdd/training/bases.py#L61) 

![training_bases_train](https://user-images.githubusercontent.com/35719368/196883978-37caf5c0-4dca-4021-bed3-2eb33a989a20.png)

![training_bases_train_data_print](https://user-images.githubusercontent.com/35719368/196884432-cb6de922-bb05-486c-9df9-88fa0c62785a.png)

We separated only the labels of the data in "self.train_loader.dataset" to see how many '0' and '1' are each present.

I thought '0' : 2875, '1' : 15 (--oe-limit 15) in 'train_loader.dataset'.
train set : '0' : 2875, '1' : 15 -> batch_size  etc -> loader
Contrary to my expectation, '0' : 1468, '1' : 1407, and these numbers change every time.
'balance_dataset' is not applied.

'test set' is being used exactly as expected.
(0 : 262, 1 : 453)

Why is this happening?

I ask you what I'm missing or mistaken about.

**Comment by liznerski** [comment]
FCDD is trained with auxiliary anomalies (e.g., Outlier Exposure). Usually, the number of normal samples and auxiliary anomalies doesn't match. Training on this would create a bias for the model that results in bad generalization. The code fixes this in one of two ways. For semi-supervision with ground-truth anomalies (i.e., `supervise_mode == 'other'`), it uses 'balance_dataset' to supersample or subsample the training anomalies so that they match the number of normal samples. For auxiliary anomalies (i.e., `supervise_mode in ['noise', 'malformed_normal', ...]`), which is what you used, the code instead randomly replaces normal samples during data loading with training anomalies with a chance of 50% (see [here](https://github.com/liznerski/fcdd/blob/master/python/fcdd/datasets/image_folder.py#L220) in _get_item_ and [here](https://github.com/liznerski/fcdd/blob/6cd39476f3232166e5ebdfbce55e3e806f082186/python/fcdd/datasets/online_supervisor.py#L87) for the OnlineSupervisor (_all_transform_)). On average, you expect to have 50% normal and 50% anomalous training labels. '0' : 1468, '1' : 1407 roughly matches this. It also explains why the number changes each time.

**Comment by SeungJunJang** [comment]
Thank you very much for your reply.

When I checked by saving the image for that part, I was wondering about the OE data being saved.
I didn't see the code that makes the 50% ratio you mentioned.

Thank you very much for your detailed reply on what I missed.
